php/class-wp-https-detection.php

```diff
@@ -0,0 +1,144 @@
+	 *
+	 * @var string
+	 */
+	const CRON_HOOK = 'check_https_support';
+
+	/**
+	 * Option name for whether HTTPS is supported.
+	 *
+	 * @var string
+	 */
+	const HTTPS_SUPPORT_OPTION_NAME = 'is_https_supported';
+
+	/**
+	 * Secret for the HTTPS detection request.
+	 *
+	 * @var string
+	 */
+	const REQUEST_SECRET = 'https_detection_secret';
+
+	/**
+	 * Query arg key for the HTTPS detection token.
+	 *
+	 * @var string
+	 */
+	const REQUEST_TOKEN_QUERY_ARG = 'https_detection_token';
+
+	/**
+	 * Inits the class.
+	 */
+	public function init() {
+		add_action( 'wp', array( $this, 'schedule_cron' ) );
+		add_action( self::CRON_HOOK, array( $this, 'update_option_https_support' ) );
+		add_filter( 'cron_request', array( $this, 'ensure_http_if_sslverify' ), PHP_INT_MAX );
+	}
+
+	/**
+	 * Schedules a cron event to check for HTTPS support.
+	 */
+	public function schedule_cron() {
+		if ( ! wp_next_scheduled( self::CRON_HOOK ) ) {
+			wp_schedule_event( time(), 'hourly', self::CRON_HOOK );
+		}
+	}
+
+	/**
+	 * Makes a request to find whether HTTPS is supported, and stores the result in an option.
+	 */
+	public function update_option_https_support() {
+		update_option( self::HTTPS_SUPPORT_OPTION_NAME, $this->is_https_supported() );
+	}
+
+	/**
+	 * Makes a request to the home URL to determine whether HTTPS is supported.
+	 *
+	 * Returns true if the URL has the scheme of HTTPS, and it has the correct token.
+	 * The token is only to verify that the request is for the same site.
+	 * Also, in wp_remote_get(), 'sslverify' is true by default.
+	 *
+	 * @return boolean Whether HTTPS is supported.
+	 */
+	public function is_https_supported() {
+		$request = wp_remote_request(
+			add_query_arg(
+				self::REQUEST_TOKEN_QUERY_ARG,
+				$this->get_token(),
+				home_url( '/', 'http' )
+			),
+			array(
+				'timeout' => 10,
+				'headers' => array(
+					'Cache-Control' => 'no-cache',
+				),
+			)
+		);
+
+		if ( is_wp_error( $request ) || ! method_exists( $request['http_response'], 'get_response_object' ) ) {
+			return false;
+		}
+
+		$response   = $request['http_response']->get_response_object();
+		$parsed_url = wp_parse_url( $response->url );
+		if ( ! $response->success || ! isset( $parsed_url['query'] ) ) {
+			return false;
+		}
+
+		$query_args = wp_parse_args( $parsed_url['query'] );
+		$is_https   = (
+			isset( $parsed_url['scheme'], $query_args[ self::REQUEST_TOKEN_QUERY_ARG ] )
+			&&
+			'https' === $parsed_url['scheme']
+			&&
+			$this->get_token() === $query_args[ self::REQUEST_TOKEN_QUERY_ARG ]
+		);
+
+		return $is_https;
+	}
+
+	/**
+	 * Gets the token, based on the secret.
+	 *
+	 * This token is not intended for security, as much as ensuring that the request is for this site.
+	 * By calling wp_hash(), it uses this site's salts, which should be different from other sites.
+	 *
+	 * @return string $token The token for the request.
+	 */
+	public function get_token() {
+		return wp_hash( self::REQUEST_SECRET );
```

Comment:
Maybe this isn't what you had in mind in this [comment](https://github.com/xwp/pwa-wp/issues/6#issue-326244070):

>For added resilience, the request should include a token which the response then hashes with a secret which can then be compared to confirm that in fact it is the same WordPress install itself that is responding, and not some other WordPress site that is configured on HTTPS.

Reply:
You could use a nonce instead like the AMP plugin is doing: https://github.com/Automattic/amp-wp/blob/a0eaa5328cdf60b22ed2d1f3e74753c6c04686e8/includes/validation/class-amp-validation-manager.php#L1410-L1435

The main thing is to just have a way to confirm that the response is actually being handled by the server we are intending. It could be a static string we look for like `WP_HTTPS_OK` but there could be a possibility that a site mentions the string and there is a false positive.

Reply:
So to summarize, the logic in `is_https_supported` could 

1. Generate a random number: `$number = wp_rand()`.
2. Make a loopback request to an HTTPS URL for the site with this `$number` added as a query var.
3. Have handler that runs at `parse_query` action that looks for this query var and then does `die( wp_hash( $query_var, 'nonce' ) )`
4. The method then takes the response body and does `wp_remote_retrieve_body( $response ) === wp_hash( $query_var, 'nonce' )`
5. If the values match, then we can confirm that the loopback request was answered by the same server. Otherwise, it did not.

Reply:
Thanks for the detailed solution. [check_https_support()](https://github.com/xwp/pwa-wp/pull/16/files#diff-1d91a59d72444acecd2a72e8baa63c2cR80) implements it mainly as you described, though the [parse_query handler](https://github.com/xwp/pwa-wp/pull/16/files#diff-1d91a59d72444acecd2a72e8baa63c2cR137) uses `wp_die()` instead of `die()`, if that's alright.